modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h

```diff
@@ -0,0 +1,68 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ADKernelValue.h"
+#include "INSADObjectTracker.h"
+
+/**
+ * Subtracts the mesh velocity from the convection term in the Navier-Stokes momentum equation
+ */
+class INSADMomentumMeshAdvection : public ADVectorKernelValue
+{
+public:
+  static InputParameters validParams();
+  static InputParameters displacementParams();
+  template <typename T>
+  static void setDisplacementParams(T & mesh_convection_obj);
+
+  INSADMomentumMeshAdvection(const InputParameters & parameters);
+
+protected:
+  virtual ADRealVectorValue precomputeQpResidual() override;
+
+  /// The strong residual for this object, computed by material classes to eliminate computation
+  /// duplication in simulations in which we have stabilization
+  const ADMaterialProperty<RealVectorValue> & _convected_mesh_strong_residual;
+};
+
+template <typename T>
+void
+INSADMomentumMeshAdvection::setDisplacementParams(T & mesh_convection_obj)
+{
+  auto check_coupled = [&](const auto & var_name)
+  {
+    if (mesh_convection_obj.coupledComponents(var_name) > 1)
+      mesh_convection_obj.paramError(
+          var_name, "Only one variable should be used for '", var_name, "'");
+    if (mesh_convection_obj.isCoupledConstant(var_name))
+      mesh_convection_obj.paramError(var_name, "Displacement variables cannot be constants");
+  };
+  check_coupled("disp_x");
+  check_coupled("disp_y");
+  check_coupled("disp_z");
+
+  if (mesh_convection_obj._tid == 0)
+  {
+    // Bypass the UserObjectInterface method because it requires a UserObjectName param which we
+    // don't need
+    auto & obj_tracker = mesh_convection_obj._fe_problem.template getUserObject<INSADObjectTracker>(
+        "ins_ad_object_tracker");
```

Comment:
should we check block restriction of that one vs this kernel

Reply:
The tracker is a general user object so it's not block restrictable. It bases its "block restriction" based off the blocks of the INSADMaterials. I already did have an error for this but the message was not very good. I will improve it